cores/esp32/USBCDC.cpp

```diff
@@ -246,14 +251,27 @@ void USBCDC::_onLineCoding(uint32_t _bit_rate, uint8_t _stop_bits, uint8_t _pari
 void USBCDC::_onRX(){
     uint8_t buf[CONFIG_TINYUSB_CDC_RX_BUFSIZE+1];
     uint32_t count = tud_cdc_n_read(itf, buf, CONFIG_TINYUSB_CDC_RX_BUFSIZE);
+
+    if(rx_queue == NULL) {
+        return;
+    }
+    if (uxQueueSpacesAvailable(rx_queue) < count) {
+        //this VTaskDelay gives, to Arduino's task, time to the CPU do its processing
+        //without it, data may be lost when the number of bytes received is higher than CDC buffer size
+        vTaskDelay(10);
+    }
     for(uint32_t i=0; i<count; i++){
-        if(rx_queue == NULL || !xQueueSend(rx_queue, buf+i, 0)){
```

Comment:
just changing the timeout here from 0 to something else would "fix" this issue. No need for the other changes (and no need for set delay). Can't stop stressing enough that you should use Arduino API when possible, not vTaskDelay

Reply:
Actually, just doing it doesn't fix it because, data may be partially copied from TUD CDC and it will return when the Arduino CDC queue is full, not sending Arduino USB CDC event to the application and dropping data.
It's actually a problem of Race Condition between USB TUD Driver (interrupt driven) and the Arduino Application.

Reply:
Regarding using Arduino ```delay()``` or FreeRTOS ```vTaskDelay()```, I think that in this case, it is not about delaying the execution, but actually a Task necessity of giving the control to a lower priority Task (Arduino Main Task) in order to allow the consumer Task to do its job. 

In the case of this issue, I think it is a lot clearer to use ```vTaskDelay()``` instead. 
The real issue here is about Race Condition and Arduino Task Starvation.
Executing ```taskYIELD()``` doesn't solve the issue because of the very high priority of ```usb_device_task```.
It must pause for a number of ```Ticks``` and allow Arduino Task to take place - considering it will consume USB CDC queue.

Please let me know your thoughts on it.

Reply:
Are we talking about the same things? Doesn't the code below do exactly what we need?
```cpp
void USBCDC::_onRX(){
    uint8_t buf[CONFIG_TINYUSB_CDC_RX_BUFSIZE+1];
    uint32_t count = tud_cdc_n_read(itf, buf, CONFIG_TINYUSB_CDC_RX_BUFSIZE);
    for(uint32_t i=0; i<count; i++){
        if(rx_queue == NULL || !xQueueSend(rx_queue, buf+i, 10)){
            count = i;
            break;
        }
    }
    if (count) {
        arduino_usb_cdc_event_data_t p;
        p.rx.len = count;
        arduino_usb_event_post(ARDUINO_USB_CDC_EVENTS, ARDUINO_USB_CDC_RX_EVENT, &p, sizeof(arduino_usb_cdc_event_data_t), portMAX_DELAY);
    }
}
```

Reply:
re `delay()`: This is Arduino code. We use our APIs when we have them. It calls vTaskDelay in the background also. Calls, implementation, anything can change, but `delay()` will always delay. Please use Arduino APIs when such are available.

Reply:
I think we should at least `log_e` here that something went wrong and bytes were dropped. It could help us in the future as well

Reply:
I have created an event ```ARDUINO_USB_CDC_OVERFLOW``` that will tell in  ```rx_overflow.dropped_bytes``` the number of bytes that were lost. Should I also add a ```log_e``` or with the event would be enough?
I also added that event to the Library examples.

Reply:
log_e is always good for us. Will print even if the user is not listening to events

Reply:
All Done - including copying queue data in setRxBufferSize()